### ISSUE TITLE
FhirPath handle variables via the context in CompiledExpressions

### ISSUE THREAD

**Issue by brianpos** [FirelyTeam/firely-net-sdk#2581]
**Is your feature request related to a problem? Please describe.**
There is currently no way to provide variables to a compiled expression.
The current model requires the explicit value to be included in the AddVar extension method.

_(Note that I've also included an option to add in collections as variables - that should be added too)_

This results in not being able to compile an expression and then provide the variables via the context for when it is to be evaluated later on.

**Describe the solution you'd like**
The variables should be able to be "registered" with the expression as a name but with no value, and then provide the value via the FhirExpressionContext.
This updated approach would then detect when variables are used that haven't been provided, and also provide values on a per execution basis without re-compilation.

**Describe alternatives you've considered**
This is a working example of how I've been able to do it without any changes to the library, but the approach could probably be included in the FhirEvaluationContext, rather than deriving it.
_(The approach was copied from the resource Resolver technique)_
``` c#
using Hl7.Fhir.ElementModel;
using Hl7.Fhir.FhirPath;
using Hl7.FhirPath;
using Hl7.FhirPath.Expressions;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using System.Collections.Generic;
using System.Linq;

namespace Test.FhirMappingLanguage
{
    [TestClass]
    public class SymbolTableVariableTests
    {

        [TestMethod]
        public void CheckVariable()
        {
            FhirPathCompiler.DefaultSymbolTable.AddFhirExtensions();
            var st = new SymbolTable(FhirPathCompiler.DefaultSymbolTable);
            st.AddVariableFromContext("current");
            st.AddVariableFromContext("previous");

            string expression = "%current | %previous";
            var compiler = new FhirPathCompiler(st);
            var expr = compiler.Compile(expression);

            var ctx = new FhirEvaluationContextWithVariables();
            ctx.SetVariable("current", ElementNode.CreateList(ElementNode.ForPrimitive(1)));
            ctx.SetVariable("previous", ElementNode.EmptyList);

            var result = expr(ElementNode.ForPrimitive(5), ctx);
            System.Diagnostics.Trace.WriteLine(string.Join(", ", result.Select(v => v.ToString())));

            ctx = new FhirEvaluationContextWithVariables();
            ctx.SetVariable("current", ElementNode.CreateList(ElementNode.ForPrimitive(1)));
            ctx.SetVariable("previous", ElementNode.CreateList(ElementNode.ForPrimitive(5)));

            result = expr(ElementNode.ForPrimitive(5), ctx);
            System.Diagnostics.Trace.WriteLine(string.Join(", ", result.Select(v => v.ToString())));
        }
    }


    public static class SymbolTableExtensions
    {
        /// <summary>
        /// Add a variable to the symbol table that contains a collection (rather than just a single value which the existing overrides provide)
        /// </summary>
        public static void AddVariable(this SymbolTable table, string name, IEnumerable<ITypedElement> value)
        {
            table.Add(name, () => { return value; });
        }

        public static void AddVariableFromContext(this SymbolTable table, string name)
        {
            table.Add(name, (EvaluationContext ctx) => { return resolveVariableFromContext(name, ctx); });
        }

        private static IEnumerable<ITypedElement> resolveVariableFromContext(string name, EvaluationContext ctx)
        {
            return ctx is FhirEvaluationContextWithVariables fctx ? fctx.GetVariable(name) : ElementNode.EmptyList;
        }
    }

    public class FhirEvaluationContextWithVariables : FhirEvaluationContext
    {
        public void SetVariable(string name, IEnumerable<ITypedElement> values)
        {
            Variables.Add(name, values);
        }

        public IEnumerable<ITypedElement> GetVariable(string name)
        {
            if (Variables.ContainsKey(name))
                return Variables[name];

            // This should really throw an error here
            return ElementNode.EmptyList;
        }

        Dictionary<string, IEnumerable<ITypedElement>> Variables { get; } = new();
    }
}
```

**Comment by ewoutkramer** [comment]
Possible duplicate of #2570.

I think if we integrate this "natively" we don't need to add things to the symbol table (as you can share a symbol table across processes and compilers, this might not cause side effects).  So, we could add an extra functionality to the "normal" resolution of `%variablename` so that after we have tried the built-in ones, and failed to find those, we check a dictionary on the FhirEvalutationContext, just like you have done above.

**Comment by ewoutkramer** [comment]
Fixed by #2570 .
